# Project #2: Automation of Concrete Mix Design (NDOT)

Group: 151-08  
Date: 2/21/2026  
Client: Nebraska Department of Transportation (NDOT)  

## Project Contex(Client summary)  

1) convert Excel logic into Python functions  
2) ask the user for inputs  
3) create a weight chart for a cubic yard  
4) evaluate four mix scenarios as verification/validation.

### 1. Convert Excel logic into Python functions  
#### 1.1 import packages

In [4]:
import pandas as pd
import numpy as np

#### 1.2 Constants

In [6]:
#Define constants
cubic_yard_ft3 = 27 #cubic feet in one cubic yard
unit_weight_water = 62.4 # lb/ft^3 (unit weight of water)

#print constants
print(f"Defined constant: 1 cubic yard = {cubic_yard_ft3} cubic feet")
print(f"Unit weight of water = {unit_weight_water} lb per cubic feet")

Defined constant: 1 cubic yard = 27 cubic feet
Unit weight of water = 62.4 lb per cubic feet


#### 1.3 Defining functions

In [9]:
# Water weight function
def calculate_water_weight_Q(cement_A,fly_ash_B,silica_fume_C, other_SCM_D, wc_ratio_E):
    total_cementitious = cement_A + fly_ash_B + silica_fume_C + other_SCM_D # calculates the amount of cementitious material in weight
    water_weight_Q = total_cementitious*wc_ratio_E # calculates water weight
    return water_weight_Q #returns water weight

#Volume of cement function
def calculate_volume_R(cement_A,sg_cement_J):
    volume_R = cement_A / (sg_cement_J * unit_weight_water) #calculates the volume
    return volume_R #returns volume of cement

#Volume of fly ash function
def calculate_volume_S(fly_ash_B, sg_fly_ash_K):
    volume_S = fly_ash_B/(sg_fly_ash_K * unit_weight_water)
    return volume_S #returns volume of fly ash

#Volume of silica fume function
def calculate_volume_T(silica_fume_C, sg_silica_fume_L):
    volume_T = silica_fume_C/(sg_silica_fume_L * unit_weight_water)
    return volume_T

#Volume of other SCM function
def calculate_volume_U(other_SCM_D, sg_other_SCM_M):
    volume_U = other_SCM_D/(sg_other_SCM_M * unit_weight_water)
    return volume_U

#Volume of air function
def calculate_air_volume_V(target_air_content_F):
    air_volume_V = (target_air_content_F/100)*cubic_yard_ft3
    return air_volume_V

#Volume of Water function
def calculate_water_volume_W(water_weight_Q):
    water_volume_W = (water_weight_Q/unit_weight_water)
    return water_volume_W

#Volume of Total Aggregates function
def calculate_total_aggregate_volume_X(volume_R, volume_S, volume_T, volume_U, volume_V, volume_W):
    aggregate_volume_X = cubic_yard_ft3 - volume_R - volume_S - volume_T - volume_U - volume_V - volume_W
    return aggregate_volume_X

#Weight of Fine Aggregates function
def calculate_fine_aggregate_Y(percent_fine_G,sg_fine_N,aggregate_volume_X):
    fine_aggregate_weight_Y = 62.4*(percent_fine_G/100)*sg_fine_N*aggregate_volume_X
    return fine_aggregate_weight_Y

#Weight of Coarse Aggregates function
def calculate_coarse_aggregate_Z(percent_coarse_H,sg_coarse_O,aggregate_volume_X):
    coarse_aggregate_weight_Z = 62.4*(percent_coarse_H/100)*sg_coarse_O*aggregate_volume_X
    return coarse_aggregate_weight_Z

#Weight of Other Aggregates function
def calculate_other_aggregate_AA(percent_other_I,sg_other_P,aggregate_volume_X):
    other_aggregate_weight_AA = 62.4*(percent_other_I/100)*sg_other_P*aggregate_volume_X
    return other_aggregate_weight_AA

### 2. Inputs and calculations

In [19]:
# NDOT Concrete Mix Design – Full Integrated Run

# Project metadata
project_no = int(input("Enter project number: "))
concrete_class = input("Enter class of concrete (e.g., 47B, 47BR): ")

# Cementitious material inputs (lb per cubic yard)
cement_A = float(input("Enter cement weight A (lb per cubic yard): "))
fly_ash_B = float(input("Enter fly ash weight B (lb per cubic yard): "))
silica_fume_C = float(input("Enter silica fume weight C (lb per cubic yard): "))
other_scm_D = float(input("Enter other SCM weight D (lb per cubic yard): "))

water_cement_ratio_E = float(input("Enter target water-cement ratio E: "))
air_content_F = float(input("Enter target air content F (%): "))

# Aggregate proportions (%)
percent_fine_G = float(input("Enter percent fine aggregate G (%): "))
percent_coarse_H = float(input("Enter percent coarse aggregate H (%): "))
percent_other_I = float(input("Enter percent other aggregate I (%): "))

# Specific gravities
sg_cement_J = float(input("Enter specific gravity of cement J: "))
sg_fly_ash_K = float(input("Enter specific gravity of fly ash K: "))
sg_silica_fume_L = float(input("Enter specific gravity of silica fume L: "))
sg_other_scm_M = float(input("Enter specific gravity of other SCM M: "))

sg_fine_N = float(input("Enter specific gravity of fine aggregate N: "))
sg_coarse_O = float(input("Enter specific gravity of coarse aggregate O: "))
sg_other_P = float(input("Enter specific gravity of other aggregate P: "))

# Calculations
Q = calculate_water_weight_Q(
    cement_A, fly_ash_B, silica_fume_C, other_scm_D, water_cement_ratio_E
)

R = calculate_volume_R(cement_A, sg_cement_J)
S = calculate_volume_S(fly_ash_B, sg_fly_ash_K)
T = calculate_volume_T(silica_fume_C, sg_silica_fume_L)
U = calculate_volume_U(other_scm_D, sg_other_scm_M)

V = calculate_air_volume_V(air_content_F)
W = calculate_water_volume_W(Q)

X = calculate_total_aggregate_volume_X(R, S, T, U, V, W)

Y = calculate_fine_aggregate_Y(percent_fine_G, sg_fine_N, X)
Z = calculate_coarse_aggregate_Z(percent_coarse_H, sg_coarse_O, X)
AA = calculate_other_aggregate_AA(percent_other_I, sg_other_P, X)

Enter project number:  4
Enter class of concrete (e.g., 47B, 47BR):  47B-3500
Enter cement weight A (lb per cubic yard):  500
Enter fly ash weight B (lb per cubic yard):  100
Enter silica fume weight C (lb per cubic yard):  0
Enter other SCM weight D (lb per cubic yard):  0
Enter target water-cement ratio E:  0.45
Enter target air content F (%):  6
Enter percent fine aggregate G (%):  39
Enter percent coarse aggregate H (%):  61
Enter percent other aggregate I (%):  0
Enter specific gravity of cement J:  3.15
Enter specific gravity of fly ash K:  2.6
Enter specific gravity of silica fume L:  2.22
Enter specific gravity of other SCM M:  2.9
Enter specific gravity of fine aggregate N:  2.62
Enter specific gravity of coarse aggregate O:  2.68
Enter specific gravity of other aggregate P:  0


### 3. Weight Chart

In [20]:
# Output: Final Weight Chart
print("\n---------------------------------------------")
print(" NDOT Concrete Mix Design – Weight Summary")
print("         (1 Cubic Yard of Concrete)")
print("---------------------------------------------")
print(f"Project Number:       {project_no}")
print(f"Class of Concrete:    {concrete_class}")
print("---------------------------------------------")
print(f"Cement (A):            {cement_A:8.1f} lb")
print(f"Fly Ash (B):           {fly_ash_B:8.1f} lb")
print(f"Silica Fume (C):       {silica_fume_C:8.1f} lb")
print(f"Other SCM (D):         {other_scm_D:8.1f} lb")
print("---------------------------------------------")
print(f"Fine Aggregate (Y):    {Y:8.0f} lb")
print(f"Coarse Aggregate (Z):  {Z:8.0f} lb")
print(f"Other Aggregate (AA):  {AA:8.0f} lb")
print("---------------------------------------------")
print(f"Water (Q):             {Q:8.0f} lb")
print("---------------------------------------------")
print("End of Mix Design Summary")


---------------------------------------------
 NDOT Concrete Mix Design – Weight Summary
         (1 Cubic Yard of Concrete)
---------------------------------------------
Project Number:       4
Class of Concrete:    47B-3500
---------------------------------------------
Cement (A):               500.0 lb
Fly Ash (B):              100.0 lb
Silica Fume (C):            0.0 lb
Other SCM (D):              0.0 lb
---------------------------------------------
Fine Aggregate (Y):        1141 lb
Coarse Aggregate (Z):      1825 lb
Other Aggregate (AA):         0 lb
---------------------------------------------
Water (Q):                  270 lb
---------------------------------------------
End of Mix Design Summary


### 4. Mix Scenarios

#### 4.1 Mix Design 1

Mix Design - 47B - 3625  
Uses - Pavement Patches, Sidewalks, Curbs, Gutters, Box Culverts, etc.  
##### Inputs

Project number - 1  
Class of Concrete - 47B  
Cement weight A - 520 lbs  
Fly Ash Weight B - 80 lbs  
Silica Fume Weight C - 0 lbs  
Other SCM Weight D - 0 lbs  
Target Water Cement Ratio E - 0.45  
Target Air Content F(%) - 5.5  
Percent Fine Aggregate G(%) - 40  
Percent Coarse Aggregate H(%) - 60  
Percent Other Aggregate I(%) - 0  
Specific Gravity of Cement J - 3.15  
Specific Gravity of Fly Ash K - 2.60  
Specific Gravity of Silica Fume L - 2.22  
Specific Gravity of Other SCM M - 2.90  
Specific Gravity of Fine Aggregates N - 2.62  
Specific Gravity of Coarse Aggregates O - 2.68  
Specific Gravity of Other Aggregates P - 0  

#### 4.2 Mix Design 2

Mix Design - 47BD - 4000  
Uses - Bridge Deck  
##### Inputs

Project number - 2  
Class of Concrete - 47BD  
Cement weight A - 450 lbs  
Fly Ash Weight B - 120 lbs  
Silica Fume Weight C - 30 lbs  
Other SCM Weight D - 0 lbs  
Target Water Cement Ratio E - 0.42  
Target Air Content F(%) - 6.0  
Percent Fine Aggregate G(%) - 43  
Percent Coarse Aggregate H(%) - 57  
Percent Other Aggregate I(%) - 0  
Specific Gravity of Cement J - 3.15  
Specific Gravity of Fly Ash K - 2.60  
Specific Gravity of Silica Fume L - 2.22  
Specific Gravity of Other SCM M - 2.90  
Specific Gravity of Fine Aggregates N - 2.62  
Specific Gravity of Coarse Aggregates O - 2.68  
Specific Gravity of Other Aggregates P - 0  

#### 4.3 Mix Design 3

Mix Design - 47B - HE3 - 3625  
Uses - Panel replacements, anything that needs a fast-setting concrete  
##### Inputs

Project number - 3  
Class of Concrete - 47BHE  
Cement weight A - 650 lbs  
Fly Ash Weight B - 0 lbs  
Silica Fume Weight C - 40 lbs  
Other SCM Weight D - 0 lbs  
Target Water Cement Ratio E - 0.40  
Target Air Content F(%) - 6.0  
Percent Fine Aggregate G(%) - 40  
Percent Coarse Aggregate H(%) - 60  
Percent Other Aggregate I(%) - 0  
Specific Gravity of Cement J - 3.15  
Specific Gravity of Fly Ash K - 2.60  
Specific Gravity of Silica Fume L - 2.22  
Specific Gravity of Other SCM M - 2.90  
Specific Gravity of Fine Aggregates N - 2.62  
Specific Gravity of Coarse Aggregates O - 2.68  
Specific Gravity of Other Aggregates P - 0  

#### Mix Design 4

Mix Design - 47B - 3500  
Uses - Highway Pavement, shoulders, etc.  
##### Inputs

Project number - 4  
Class of Concrete - 47B  
Cement weight A - 500 lbs  
Fly Ash Weight B - 100 lbs  
Silica Fume Weight C - 0  
Other SCM Weight D - 0  
Target Water Cement Ratio E - 0.45  
Target Air Content F(%) - 6.0  
Percent Fine Aggregate G(%) - 39  
Percent Coarse Aggregate H(%) - 61  
Percent Other Aggregate I(%) - 0  
Specific Gravity of Cement J - 3.15  
Specific Gravity of Fly Ash K - 2.60  
Specific Gravity of Silica Fume L - 2.22  
Specific Gravity of Other SCM M - 2.90  
Specific Gravity of Fine Aggregates N - 2.62  
Specific Gravity of Coarse Aggregates O - 2.68  
Specific Gravity of Other Aggregates P - 0  